### DISCUSSION TITLE
Declare variables using mutation functions

### DISCUSSION THREAD

**Discussion by cchannon** [microsoft/Power-Fx#827] [Q&A] [answered]
I am tying myself into knots trying to identify and parse all mutation functions prior to Eval so I can avoid the ```"Name isn't valid. 'XXXX' isn't recognized"``` error. And it just occurred to me that the Canvas App authoring experience allows for seamless inline declaration - is there just a flag or a feature of some kind I need to enable that will tell engine.Eval to permit inline declaration by mutation functions?

**Comment by anderson-joyle** [comment]
Can you please give me an example of what you are experiencing?

**Reply by cchannon** [reply]
Sure. So, let's say I instantiate a new engine, and I don't pass it any variables as part of the construction. Now I pass Set(myvar,1); straight into engine.Eval(). 

This will cause the engine to throw an error because myvar hasn't already been declared with some other method like .UpdateVariable()

OK, that makes sense - it is important to ensure the variable _should_ exist and that we are being careful about types. The problem is, sometimes I _want_ to be less rigorous with my users. In Canvas Apps, for example, there is no need to go to the Variables section and declare it before you start putting it into Behavior properties - you just add it to a Behavior property and the evaluation engine is smart enough to declare it for you.

However, when you get down into the nuts and bolts of making that work through brute force, it is very challenging. At the surface of a .Check result there is a list of all the idents and that is a GREAT start because it at least gives me a shortcut to find out if something is being used in a mutation function without already being declared. BUT, it doesn't have a corresponding list of inferred types, so if I find one that hasn't been declared, now we need to dig down into the object to whatever depth is necessary to _find_ the mutation function, _parse_ it, _infer_ the type, then _declare_ the var, then Eval the statement.

This Q&A isn't about how do I do that - it is because I realized once I got there that you must have some better way, and I just don't know about it yet.

**Comment by MikeStall** [comment, marked as answer]
Unfortunately, there isn't a helper function to do this yet.  So you will need to manually handle it like you outlined above. 
Part of the reason is that the analysis can get very complex in "whole document" scenarios. For example, `Set(x,y); Set(y,z); Set(z,x)`  would need to detect the chain and circular reference.  And this could potentially spread across multiple expressions. 

Note that this can apply to Set, Collect, ClearCollect and possibly other functions in the future. 

curious - do you need Set, or  can you use Named Formulas?  See Engine.SetFormula.

**Reply by cchannon** [reply]
Ah. Well, yeah, the challenges are real, you are correct.

Well, I would _prefer_ to allow users to use Set - it is a familiar function and telling them, "This is PFX, but you can't Set," would make for a sub-par UX.

That said, would Named Formulas handle this problem differently? I mean, is there procedurally a difference in my Eval headaches whether the expr is 
```
If(true,If(true,If(true,Set(myvar,"a")),Set(anothervar,"b")),Set(someothervar,3))
``` 
or
```
myformula = If(true,If(true,If(true,Set(myvar,"a")),Set(anothervar,"b")),Set(someothervar,3))
```

It seems to me like I need to do the same hunting either way.

**Comment by anderson-joyle** [comment]
Another possible solution would be to extend Engine/RecalEngine class into your own Engine to identify and address any missing variable.

**Comment by cchannon** [comment]
OK, so I am going to close this one out for now. It seems to me that at this point in the language evolution the only approach I can take that doesn't send me down endless rabbit-holes is just to force my users to declare Context and Commands in separate properties, and that there is--for the moment--no practical way for me to offer users a single scripting window where they can seamlessly write Commands and have Variables implicitly declared.

**Reply by MikeStall** [reply]
Here's a demo of how to make Set() declare variables.  It's still painful, but better than a regex. 

https://github.com/microsoft/Power-Fx/blob/a45e82aa2abc939fbec783465df68b65864b117a/src/tools/Repl/Program.cs#L81

This technique is used in PR #926 to start having "stateful repl" tests.

**Reply by cchannon** [reply]
Beautiful. Expect a PR on the Samples side in the next couple of days. A REPL like this would have saved me tons of time instead of the one that's currently in there!

**Reply by MikeStall** [reply]
We copied the REPL into Power Fx core - and it's already included in that version :) 

https://github.com/microsoft/Power-Fx/blob/main/src/tools/Repl/Program.cs

We found having a copy of the repl directly in core is easier to work with.
